### DISCUSSION TITLE
Export to UWP relevant Bitmap formats

### DISCUSSION THREAD

**Discussion by jhwheuer** [SixLabors/ImageSharp#1437] [Ideas]
Since shoveling bitmap arrays between file and memory is a bit cumbersome (not to say painful with the UWP disk-IO framework), I would like to be able to **export the result of an Image manipulation** to a [`Windows.Graphics.Imaging.SoftwareBitmap`](https://docs.microsoft.com/en-us/uwp/api/windows.graphics.imaging.softwarebitmap?view=winrt-19041) or a [`Windows.UI.Xaml.Media.Imaging.WriteableBitmap`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.imaging.writeablebitmap?view=winrt-19041).

Any idea how to do that, or where to put a hook?
I could envision an Encoder that does something with this kind of approach... [https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.imaging.writeablebitmap.pixelbuffer?view=winrt-19041](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.imaging.writeablebitmap.pixelbuffer?view=winrt-19041)

**Comment by dlemstra** [comment]
Another hook could be an extension method that does something similar to the approach that I use in Magick.NET: https://github.com/dlemstra/Magick.NET/blob/c7c98835b017277820884b574823f6da049db5ef/src/Magick.NET.SystemWindowsMedia/IMagickImageExtentions.cs. This copies the pixels to from the image to a `BitmapSource`. You could then do something like `var bitmap = image.ToWriteableBitmap()`

**Reply by jhwheuer** [reply]
THANK YOU. 
That is one of the areas that I do not understand (3 hours into the framework)... how to I get my hands on a copy of the actual pixels in a frame...

**Reply by JimBobSquarePants** [reply]
You're looking for [`Image<TPixel>.TryGetSinglePixelSpan(out Span<TPixel>)`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Image-1.html#SixLabors_ImageSharp_Image_1_TryGetSinglePixelSpan_System_Span__0___)

Note, that very large images are not contiguous in memory, but for images of size displayable on screen (below ~4k*4k) we plan to guarantee that the memory group will contain a single memory buffer, because of the GPU interop use-cases. See also: #1258

**Comment by antonfirsov** [comment]
@jhwheuer also check out the usage example in #1419

This is about 3rd UWP-interop request I see in the past month. If you manage to figure it out, any chance you can post a code example?

@Sergio0694 any interest to contribute such examples to https://github.com/SixLabors/Samples ?

**Reply by jhwheuer** [reply]
Hi, sorry for the tardy response. I am currently creating a universal Wrapper for my UWP app to run arbitrary operations in sequence for my upcoming app http://www.just-snips.com/. Right now I am dumping the processing results back into a temp file...

I will take the `TryGetSinglePixelSpan` approach and see how much belly ache UWP is going to have with that. Anything I am processing needs to be shown on a display, so I might just hardcode a threshold and return a StorageFile rather than a xBitmap object...



The library does seem to have hickups when processing **animated GIFs**... anything you might share there??? Mutate throws exceptions when changing Brightness etc... Once I know more, I will post an issue, but maybe you guys can just point me in the right direction? (AKA Don't Do Dat) :-)

Thanks again
 J

**Reply by tocsoft** [reply]
nope raise the issue it should work fine... might be related to number of frames and frame size when we try and processes them but it also could be anything so best bet is to just raise with as much info as possible and to not forget to attach the image cause issues (is a zip is usually best as github can end up processing uploads otherwise)

**Reply by Sergio0694** [reply]
> any interest to contribute such examples to https://github.com/SixLabors/Samples ?

@antonfirsov Sure thing, will make a PR for the sample once that new API is merged and I have some time! 😊
